src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsWriter.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec.KNN990Codec;
+
+import lombok.RequiredArgsConstructor;
+import org.apache.lucene.codecs.KnnFieldVectorsWriter;
+import org.apache.lucene.codecs.KnnVectorsWriter;
+import org.apache.lucene.codecs.hnsw.FlatVectorsWriter;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.MergeState;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.index.Sorter;
+import org.apache.lucene.util.IOUtils;
+import org.apache.lucene.util.RamUsageEstimator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A KNNVectorsWriter class for writing the vector data strcutures and flat vectors for Native Engines.
+ */
+@RequiredArgsConstructor
+public class NativeEngines990KnnVectorsWriter extends KnnVectorsWriter {
+    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(NativeEngines990KnnVectorsWriter.class);
+    private final SegmentWriteState segmentWriteState;
+    private final FlatVectorsWriter flatVectorsWriter;
+    private final List<NativeEngineFieldVectorsWriter<?>> fields = new ArrayList<>();
+    private boolean finished;
+
+    /**
+     * Add new field for indexing.
+     * In Lucene, we use single file for all the vector fields so here we need to see how we are going to make things
+     * work.
+     * @param fieldInfo {@link FieldInfo}
+     */
+    @Override
+    public KnnFieldVectorsWriter<?> addField(final FieldInfo fieldInfo) throws IOException {
```

Comment:
Can't we avoid using `?`. It is hard to know what the `?` can be and what we can do with `?`.

Reply:
No we cannot avoid. If I avoid it and lets say make it just KnnFieldVectorsWriter, then warning comes up saying Raw usage of Parametrized class KnnFieldVectorsWriter. Hence we have to use `?`

Reply:
That is one place where wrapping might be good over primitive array.

Reply:
conversation continued here: https://github.com/opensearch-project/k-NN/pull/1855#discussion_r1692070831